include/pagmo/algorithms/gaco.hpp

```diff
@@ -0,0 +1,995 @@
+    Software Foundation; either version 3 of the License, or (at your
+    option) any later version.
+or
+  * the GNU General Public License as published by the Free Software
+    Foundation; either version 3 of the License, or (at your option) any
+    later version.
+or both in parallel, as here.
+The PaGMO library is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+You should have received copies of the GNU General Public License and the
+GNU Lesser General Public License along with the PaGMO library.  If not,
+see https://www.gnu.org/licenses/. */
+
+#ifndef PAGMO_ALGORITHMS_GACO_HPP
+#define PAGMO_ALGORITHMS_GACO_HPP
+
+#include <algorithm> // std::shuffle, std::transform
+#include <iomanip>
+#include <numeric> // std::iota, std::inner_product
+#include <random>
+#include <sstream> //std::osstringstream
+#include <string>
+#include <tuple>
+#include <utility> //std::swap
+#include <vector>
+
+#include <pagmo/algorithm.hpp> // needed for the cereal macro
+#include <pagmo/exceptions.hpp>
+#include <pagmo/io.hpp>
+#include <pagmo/population.hpp>
+#include <pagmo/problem.hpp>
+#include <pagmo/problems/decompose.hpp>
+#include <pagmo/rng.hpp>
+#include <pagmo/utils/generic.hpp>
+#include <pagmo/utils/generic.hpp>         // uniform_real_from_range
+#include <pagmo/utils/multi_objective.hpp> // crowding_distance, etc..
+
+namespace pagmo
+{
+/// Extended ACO
+/**
+ * \image html ACO.jpg "The ACO flowchart" width=3cm [TO BE ADDED]
+ * ACO is inspired by the natural mechanism with which real ant colonies forage food.
+ * This algorithm has shown promising results in many trajectory optimization problems.
+ * The first appearance of the algorithm happened in Dr. Marco Dorigo's thesis, in 1992.
+ * ACO generates future generations of ants by using the a multi-kernel gaussian distribution
+ * based on three parameters (i.e., pheromone values) which are computed depending on the quality
+ * of each previous solution. The solutions are ranked through an oracle penalty method.
+ *
+ *
+ * The version implemented in pagmo can be applied to box-bounded multiple-objective optimization.
+ *
+ * See:  M. Schlueter, et al. (2009). Extended ant colony optimization for non-convex
+ * mixed integer non-linear programming. Computers & Operations Research.
+ */
+class g_aco
+{
+public:
+    /// Single entry of the log (gen, fevals, ideal_point)
+    typedef std::tuple<unsigned, double, double, double, double, double, double> log_line_type;
+    /// The log
+    typedef std::vector<log_line_type> log_type;
+
+    /**
+     * Constructs the ACO user defined algorithm for single and multi-objective optimization.
+     *
+     *
+     *  @param[in] gen Generations: number of generations to evolve.
+     * @param[in] acc Accuracy parameter: for inequality and equality constraints .
```

Comment:
I did not update the description of this parameter. I basically introduced this acc to let the user define the minimum distance between penalty values for a solution to be accepted. I will update the description!

Reply:
Is this not the same meaning as the ```c_tol``` of the pagmo::problem? Why not using it then? We already ask the user to define a tolerance on the constraint satisfaction.

If its not the same, then indeed the documentation must be clearly addressing the semantic difference between the two.

Reply:
I was planning to use c_tol for when computing the inequality and equality constraints violations and then to use another tolerance (this acc) for maintaining a certain distance between penalties. 
For instance, if, in mathematical terms, g_e(X)=0, then, in the code, I would have considered g_e(X)=0 also if g_e(X)<=acc. And this would have been done inside the penalty computation. Afterward, I would have also applied a minimum distance between the sorted penalties: so that if a solution is not at least |pen_i - pen_{i-1}|>=acc, then the solution is not accepted.

However, now that you pointed out this, and since these two parameters are somehow related to the same aspects, it is probably more efficient and simpler to directly use c_tol for the penalties, and leave the equality and inequality constraint violations computation in mathematical terms.

Reply:
Did you consider using detail::test_eq_constraints and detail::test_ineq_constraints ? they are in ```pagmo/utils/constrained.hpp``` maybe they can be of use for your case.

Reply:
You can pass to these fnctions c_tol .... and get the L2 violation discounted by c_tol

Reply:
Cool! I did not notice them! Thanks!